refactor: add new module for stickers

jrainville opened the pull request.

saledjenic commented on status/statusgo_backend_new/stickers.nim:
```diff
@@ -0,0 +1,74 @@
+import json, stint, web3/[ethtypes, conversions], strutils, tables, sequtils
+import
+  ../types/[setting],
```
@jrainville all lines here from 3 to 11 rely on the "old" `status-lib` code, and that's not the point here. `statusgo_backend_new` part is created with an idea to have here only new logic (actually calls to `status-go`) and nothing else. Cause at the end of refactoring phase we will remove everything else and only files under `statusgo_backend_new` will make `status-lib`.

saledjenic commented on status/statusgo_backend_new/stickers.nim:
```diff
@@ -0,0 +1,74 @@
+  approveAndCall = ApproveAndCall[100](to: stickerMktContract.address, value: price, data: DynamicBytes[100].fromHex(buyTxAbiEncoded))
+  transactions.buildTokenTransaction(address, sntContract.address, gas, gasPrice, isEIP1559Enabled, maxPriorityFeePerGas, maxFeePerGas)
+
+proc getInstalledStickerPacks*(): JsonNode =
```
@jrainville it's not part of the lib to this parsing, it's only here as a bridge for communication with `status-go`.
All methods from the `status-lib` should have the same return type `RpcResponse[JsonNode] {.raises: [Exception].}` and that will be handled in appropriate service on the `status-desktop` side.

saledjenic commented on status/statusgo_backend_new/stickers.nim:
```diff
@@ -0,0 +1,74 @@
+  if success:
+    trackPendingTransaction(result, address, $sntContract.address, PendingTransactionType.BuyStickerPack, $packId)
+
+proc saveRecentStickers*(stickers: seq[string]) =
```
@jrainville all methods from the status-lib should have the same return type `RpcResponse[JsonNode] {.raises: [Exception].}` regardless we're using that response or not, `status-lib` doesn't care about that, the decision about that will be made in appropriate service on the `status-desktop` side.

saledjenic reviewed (changes requested):
@jrainville you're calling methods from the "old" status-lib

jrainville commented:
That was a mistake. Oops.

jrainville commented:
@saledjenic I addressed most of your comments. There is still `buyPack` that doesn't do the `RPCResponse`, but it's because it's pretty complex and it also doesn't use `callPrivateRPC`, is uses `sendTransaction`.

Otherwise, there are still a lot of imports that you said we should get rid of, but again, it's really complex I think.
Some of them are just Utils that are still useful even in a world where we only call status_go, like the contracts function. If we get rid of them, we have to duplicate a lot of complex code.
If you want, maybe you can tell me which components/files we want to get rid of first and I can focus on that.

saledjenic reviewed (changes requested):
@jrainville if you agree, let's do the things immediately the way they should be. At the moment it will be maybe a bit more work, but we're finalizing them forever.

This `stickers.nim` file as a part of `status-lib` knows more than it should know and thus it is involved in some logic (decision) things, and we don't want that, we just want `status-lib` to be here just as a bridge for communication with the `status-go`. At the end all logic will be either in the services of the desktop app or in the `status-go`.

As a good example, which contained complex functions before refactor, you may see account file [here](https://github.com/status-im/status-lib/blob/refactoring/base_bc/status/statusgo_backend_new/accounts.nim). Each method there contains just a call to `status-go` and forwards response back as `RpcResponse[JsonNode]`, nothing else.

Now let's analyze the file here:
- getBalance
  - `findContract` method is not part of the `status-lib` since it searches through the all contracts in order to find the right one, so we will deal with it on the `status-desktop` side 
  - `status-lib` shouldn't know about `BalanceOf`
  - the same for `encodeAbi `
  this method should look like `proc getBalance*(address: Address, encodedCotractMethods: string): RpcResponse {.raises: [Exception].} =`
so we're fetching balance based on address and encoded contract methods, and appropriate app service will take care how to prepare and send correctly those data. Main app knows about contracts, and know how to encode them using `encodeAbi`

- tokenOfOwnerByIndex
  - very similar as getBalance, the rest is resposibility of the main app (service)
  `proc tokenOfOwnerByIndex*(address: Address, encodedCotractMethods: string): RpcResponse {.raises: [Exception].} =`

- getPackIdFromTokenId
  - the same as above

- buyPack
  - in general the same, but line 30 here `transactions.buildTokenTransaction(....)` doesn't have anything with status lib, that also will be part of the service, and as I see instead of  calling `buyPack` and `buildTransaction` from it, you will need to add 
 ```
proc trackPendingTransaction*(hash: string, fromAddress: string, toAddress: string, transactionType: string, data: string) =
  let payload = %* [{"hash": hash, "from": fromAddress, "to": toAddress, "type": transactionType, "additionalData": data, 
  "data": "",  "value": 0, "timestamp": 0, "gasPrice": 0, "gasLimit": 0}]
  result = callPrivateRPC("wallet_storePendingTransaction", payload)

  ```

So work within `statusgo_backend_new` folder like it's the only here, cause at the end only files from there will compose `status-lib`.

I hope this make sense for you? And if you need anything regarding this let me know, maybe I can help.

jrainville commented:
@saledjenic I think I can make you proud. The stickers file is now bare of any imports save for the calls.
You can re-review again.
Thanks

saledjenic commented:
@jrainville it's perfect, thank you.

If it's not a problem, please do the following:
- (this is not mandatory it's up to you) in `status/statusgo_backend_new/eth.nim`, method name `call` is a bit vague what does it do, I would suggest new name like `doEthCall` or `fetchEthData` or something like that, but as I said not mandatory, as you wish
- squash your commits to a single one
- rebase them to `refactoring/base_bc`

jrainville committed:
refactor: add new module for stickers
